### PR TITLE
[FLAVA] Separate out text and image encoders

### PR DESCRIPTION
**Pull request by ankitade**
Separate out the encoders into their own module without ay logic changes (except fixing 2 minor bugs, see annotations by me) and add tests

Test plan:
pytest

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102
* #96

### CODE REVIEW

File: torchmultimodal/models/flava/flava_model.py [hunk 1]
```diff
@@ -592,439 +499,6 @@ def forward(
...
-        """
-
-        npatch = embeddings.shape[1] - 1
-        n = self.position_embeddings.shape[1] - 1
-        if npatch == n and height == width:
-            return self.position_embeddings
-        class_pos_embed = self.position_embeddings[:, 0]
-        patch_pos_embed = self.position_embeddings[:, 1:]
-        dim = embeddings.shape[-1]
-        h0 = height // self.config.patch_size
```
Comment (ankitade) [inline torchmultimodal/models/flava/flava_model.py:677]: changed to self.patch_embedding.patch_size

File: torchmultimodal/models/flava/flava_model.py [hunk 2]
```diff
@@ -592,439 +499,6 @@ def forward(
...
-            pixel_values=pixel_values,
-            image_patches_mask=image_patches_mask,
-            attention_mask=attention_mask,
-        )
-        return FLAVATransformerOutput(
-            last_hidden_state=output.last_hidden_state,
-            pooler_output=output.pooler_output,
-            hidden_states=output.hidden_states,
-            attentions=output.attentions,
-            image_labels=image_labels,
```
Comment (ankitade) [inline torchmultimodal/models/flava/flava_model.py:825]: removed this, image_labels is not a field

**Comment by ankitade** [comment]
@ankitade has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37370959).

File: torchmultimodal/models/flava/flava_text_encoder.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            attention_mask[input_ids == self.pad_token_id] = 0
+        if token_type_ids is None:
+            token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=device)
+
+        # We can provide a self-attention mask of dimensions
+        # [batch_size, from_seq_length, to_seq_length]
+        # ourselves in which case we just need to make it broadcastable to all heads.
+        extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(
+            attention_mask, input_shape, device
+        )
```
Comment (ebsmothers) [inline torchmultimodal/models/flava/flava_text_encoder.py:194]: This doesn't need to be a class method. Katrina is adding it as a util in #99

File: test/models/flava/test_flava_image_encoder.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                        [-0.9840, 0.9840],
+                        [-0.9840, 0.9840],
+                        [-0.9840, 0.9840],
+                    ],
+                ]
+            ),
+            atol=1e-4,
+            rtol=0,
+        )
+        assert_expected(out.pooler_output, out.last_hidden_state)
```
[the remark this answers is not in the document]
Comment (ankitade) [reply test/models/flava/test_flava_image_encoder.py:95]: isnt the transitive thing actually making it clear which values should line up

File: torchmultimodal/models/flava/flava_image_encoder.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+class PatchEmbeddings(nn.Module):
+    """
+    Image to Patch Embedding.
+    """
+
+    def __init__(self, image_size=224, patch_size=16, num_channels=3, embed_dim=768):
```
[the remark this answers is not in the document]
Comment (ankitade) [reply torchmultimodal/models/flava/flava_image_encoder.py:37]: typing will get handled separately

File: torchmultimodal/models/flava/flava_image_encoder.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        # see discussion at https://github.com/facebookresearch/dino/issues/8
+        h0, w0 = h0 + 0.1, w0 + 0.1
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed.reshape(
+                1, int(math.sqrt(n)), int(math.sqrt(n)), dim
+            ).permute(0, 3, 1, 2),
+            scale_factor=(h0 / math.sqrt(n), w0 / math.sqrt(n)),
+            mode="bicubic",
+            align_corners=False,
+        )
```
[the remark this answers is not in the document]
Comment (ankitade) [reply torchmultimodal/models/flava/flava_image_encoder.py:124]: Generally I want to avoid touching core logic as part of this refactor. I have a feeling some of the image encoder is going to get deleted in the end

File: torchmultimodal/models/flava/flava_text_encoder.py (hunk 3 above)
Comment (ankitade) [reply torchmultimodal/models/flava/flava_text_encoder.py:194]: actually this might need some more thought. do we want to handle attention mask HF style?

Comment (apsdehal) [reply torchmultimodal/models/flava/flava_text_encoder.py:194]: I would not touch the FLAVA's core logic in general. Feel free to refactor stuff around the logic but it would be good to avoid touching the logic itself.

**Comment by ankitade** [comment]
general comment, not going to introduce deeper refactoring changes in this PR since first time we are adding tests for the encoders (will get handled as part of unification / cleanup)

File: test/models/flava/test_flava_image_encoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/models/flava/test_flava_image_encoder.py:95]: Sure, just personal preference I guess

File: torchmultimodal/models/flava/flava_text_encoder.py (hunk 3 above)
Comment (ankitade) [reply torchmultimodal/models/flava/flava_text_encoder.py:194]: I think we discussed this @apsdehal, it should be fine as long as tests are added, they pass and ckpt is kept in sync. 

In general we need to refactor for somethings. examples: the projection being part of only pretraining model (but its needed for zero shot) or trying to use a common implementation of transformers. Will add you to all the PRs so we can address any concerns you have.
